File: engines/unity/flutter_plugin/android/src/main/kotlin/org/cytoid/gamecore/CytoidGameCorePlugin.kt
package org.cytoid.gamecore

import android.app.Activity
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

class CytoidGameCorePlugin : FlutterPlugin, ActivityAware, MethodChannel.MethodCallHandler {
    private var methodChannel: MethodChannel? = null
    private var eventChannel: EventChannel? = null
    private var bridge: CytoidGameCoreBridge? = null
    private var activity: Activity? = null

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel = MethodChannel(binding.binaryMessenger, METHOD_CHANNEL_NAME).also {
            it.setMethodCallHandler(this)
        }
        eventChannel = EventChannel(binding.binaryMessenger, EVENT_CHANNEL_NAME)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel?.setMethodCallHandler(null)
        methodChannel = null
        eventChannel?.setStreamHandler(null)
        eventChannel = null
        bridge?.dispose()
        bridge = null
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
        attachBridgeIfReady()
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activity = null
        bridge?.detachActivity()
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.activity
        attachBridgeIfReady()
    }

    override fun onDetachedFromActivity() {
        activity = null
        bridge?.detachActivity()
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        val currentBridge = bridge
        if (currentBridge == null) {
            result.error("no_activity", "Game core plugin is not attached to an Activity.", null)
            return
        }

        when (call.method) {
            "send" -> {
                val json = call.arguments as? String
                if (json == null) {
                    result.error("invalid_argument", "Expected envelope JSON string for send.", null)
                    return
                }
                currentBridge.onOutboundMessage(json)
                result.success(null)
            }
            "getEngineMode" -> result.success(currentBridge.engineMode)
            "queryRuntimeStatus" -> result.success(currentBridge.runtimeStatus())
            "ensureRuntimeStarted" -> {
                currentBridge.ensureRuntimeStarted()
                result.success(null)
            }
            "showGameSurface" -> currentBridge.showGameSurface(result)
            "hideGameSurface" -> {
                currentBridge.hideGameSurface()
                result.success(null)
            }
            else -> result.notImplemented()
        }
    }

    private fun attachBridgeIfReady() {
        val currentActivity = activity ?: return
        val currentEventChannel = eventChannel ?: return
        val currentBridge = CytoidGameCoreBridge.getOrCreate(currentActivity).also {
            it.attachActivity(currentActivity)
        }
        bridge = currentBridge
        currentEventChannel.setStreamHandler(currentBridge)
    }

    companion object {
        private const val METHOD_CHANNEL_NAME = "cytoid/game_core"
        private const val EVENT_CHANNEL_NAME = "cytoid/game_core/events"
    }
}

File: engines/unity/flutter_plugin/ios/cytoid_game_core/Sources/cytoid_game_core/CytoidGameCorePlugin.swift
import Flutter
import UIKit

public class CytoidGameCorePlugin: NSObject, FlutterPlugin {
  private static let methodChannelName = "cytoid/game_core"
  private static let eventChannelName = "cytoid/game_core/events"
  private static let waitForReadyChannelName = "cytoid_game_core/waitForReady"

  private let bridge = CytoidGameCoreBridge()

  public static func register(with registrar: FlutterPluginRegistrar) {
    let instance = CytoidGameCorePlugin()

    let methodChannel = FlutterMethodChannel(
      name: methodChannelName,
      binaryMessenger: registrar.messenger()
    )
    registrar.addMethodCallDelegate(instance, channel: methodChannel)

    let waitForReadyChannel = FlutterMethodChannel(
      name: waitForReadyChannelName,
      binaryMessenger: registrar.messenger()
    )
    registrar.addMethodCallDelegate(instance, channel: waitForReadyChannel)

    let eventChannel = FlutterEventChannel(
      name: eventChannelName,
      binaryMessenger: registrar.messenger()
    )
    eventChannel.setStreamHandler(instance.bridge)

    UnityGameCoreCallback.register(instance.bridge)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "send":
      guard let json = call.arguments as? String else {
        result(
          FlutterError(
            code: "invalid_argument",
            message: "Expected envelope JSON string for send.",
            details: nil
          )
        )
        return
      }
      bridge.onOutboundMessage(json)
      result(nil)
    case "getEngineMode":
      result(bridge.engineMode)
    case "queryRuntimeStatus":
      result(bridge.runtimeStatus())
    case "ensureRuntimeStarted":
      bridge.ensureRuntimeStarted()
      result(nil)
    case "showGameSurface":
      bridge.showGameSurface(result: result)
    case "hideGameSurface":
      bridge.hideGameSurface()
      result(nil)
    case "waitForReady":
      // Optional per-call timeout in seconds. Falls back to the v2
      // recommended default (30s) when omitted.
      let timeout: TimeInterval
      if let value = call.arguments as? Double {
        timeout = value
      } else if let value = call.arguments as? Int {
        timeout = TimeInterval(value)
      } else {
        timeout = CytoidGameCoreBridge.waitForReadyDefaultTimeout
      }
      Task {
        do {
          try await bridge.waitForReady(timeout: timeout)
          result(nil)
        } catch let error as CytoidGameCoreBridge.WaitForReadyError {
          switch error {
          case .timeout:
            result(FlutterError(code: "waitForReadyTimeout", message: "Runtime did not reach ready state within \(timeout)s.", details: nil))
          case .alreadyFailed(let code):
            result(FlutterError(code: code, message: "Runtime is in a failed state.", details: nil))
          }
        } catch {
          result(FlutterError(code: "waitForReadyFailed", message: "\(error)", details: nil))
        }
      }
    default:
      result(FlutterMethodNotImplemented)
    }
  }
}
